### ISSUE TITLE
Preparation for Java 8 OCA 1Z0 808 exam

### ISSUE THREAD

**Issue by monami555** [monami555/monami555.github.io#30]

**Comment by VillieldrVuur** [comment]
Hi, 

Thanks for the awesome rundown of surprises. I'm sure this is going to help a lot of people out.

Could you please clarify the content about list.toArray(); ?

I created a test application that changes the array and the List, but I don't see any changes in the contents of the related data structure.

Also, I don't get an exception when I change the sizes of the array or List.

**Comment by monami555** [comment]
Hello VillieldrVuur,

Indeed, you spotted a mistake :) What I had in mind while writing it, was `Arrays.asList()`, not `list.toArray()` :). And I just checked that you cannot pass the array to `Arrays.asList()`, and the toArray() method indeed creates a new array - therefore the comment about changes in one reflected in another one does not make sense. The exception comment is still valid though - you can get it by calling `list.remove(0);`, after the list has been created by `Arrays.asList()`.

Thanks for spotting that, I will correct it in the text :)

Monika

**Comment by VillieldrVuur** [comment]
Another point:

Under the section "On the self assesment test (40% correct)", you say "printing list without .toString() prints only the object ref of course"

Are you sure of this?
I wrote a small app with the code:
List<String> stringList = Arrays.asList("1", "2", "3", "4");
System.out.println(stringList);
System.out.println(stringList.toString());

And, the output is:
[1, 2, 3, 4]
[1, 2, 3, 4]

Sorry for the constant comments. I'm going through this as I study for the exam :)

**Comment by monami555** [comment]
Hello, no problem, I'm glad someone makes use of it :)

If you look inside implementation of System.out.println() you will see:

public void println(Object x) {
    String s = String.valueOf(x);

which calls toString(), so in the end those two lines do exactly the same.. I guess I meant something else than System.out.println(), though now spent some time trying to figure it out and I cannot think of any way of printing something without having it converted to String. I guess I will just remove that line, please disregard it :) Ah, and remember not to mix up printing an array with printing a list :P

**Comment by VillieldrVuur** [comment]
Ja, I think you meant printing an Array, although it will print the object reference regardless. Btw, would you mind if, when I pass the exam, I link this page for other people? It's proving to be extremely helpful with regards to preparing for the exam. It's weird how I had to search for a while to find any site that could provide anyone's experiences with the exam :(

**Comment by monami555** [comment]
Yes sure I don't mind, the site is public :) Sharing it is even recommended ;)

**Comment by VillieldrVuur** [comment]
Awesome, I will do so. Although, I will definitely be adding as much as I can as I go along ;)

**Comment by VillieldrVuur** [comment]
Hey, it's me again! I've just summarised the contents of the book. Would it be possible for you to provide detail as to how you studied for the exam?

As a Java developer, I feel confident about this exam, but I KNOW there will be tricks that will appear. What did you do to make sure that you are prepared for the exam apart from reading the materials?

**Comment by monami555** [comment]
Hi!

I guess that by now it's a bit late to answer, hehe ;) How did it go?

Basically I read the book, did the tests from the book (in real IDE) and wrote this blog post.. to be honest I was just so tired with studying that I thought "ok if this is not enough to pass then I don't want it, as it's not worth any more of my time", I think this is a good indicator that one is ready (to pass or to fail) :P

**Comment by visioncoder** [comment]
Hi monami555,

Great stuff! Very thoughtful of you to put this up for those following your footsteps :)

Comparing your OCA chapter scores with what you scored in the actual exam, how would you rate the difficulty level of the real test?

**Comment by monami555** [comment]
Hi thanks :) I do not remember it that well anymore but I think that the book was quite realistically reflecting the difficulty of the exam. For sure the parts where I scored less with the book I have repeated again (40% with the book is for sure not enough) and tried to memorise the weak points, as well as learn what to pay attention to in how the questions are formulated (bastards!). I would say that the book was for sure enough.

Previously I did 2 Spring exams and for them I felt I needed to know 100% to fit in the error margin due to some questions being not precise enough. I remember with Java it was not that tough, but still I had the habit of aiming at 100%, and with that I do not remember any impression that it was hard.

**Comment by ldejonghe** [comment]
Nice summary, well done.
However, beware of your statement : 

> it is possible to override a protected method with a public; if the method is not protected but private, there is no compiler error, and the two are completely separate;

Note that if you code correctly, when overriding a method, you **should use** the annotation Override.
It has been made especially for the purpose detecting anomalies, cfr. misspelling the function in extending class.  (or a function being private in the parent class)  
 `@Override`

A result, you will get an error in case you override a method which has been declared private in the parent , since a private function in a parent class is not seen by the extended class(only that specific class can use it) . So, it is perfectly normal you then have 2 separate functions. 
But you will have compiler errors if you use the Override annotation (and you should always use it when you have the intention to override, to have a check on the function finding a match in the parent)

**Comment by naikel** [comment]
@ldejonghe even though you're right, this is an article for preparation to the OCA Java 8 exam, and all annotations are way out of scope for this certification.  Annotations are covered in the OCP exam. So basically in this exam, they test you so you can identify the inheritance problems, not your compiler.

**Comment by cyberglad** [comment]
Exam is full of errors, from 70 question I counted about 6 or 7 errors.  Wrong results, blocks that actually compile (when it says it wouldn't), unreachable code, let alone syntax erros. They need to revise it
